src/mappings/entities/mapToken.ts

```diff
@@ -0,0 +1,165 @@
+import BN from 'bn.js';
+import { SubstrateExtrinsic } from '@subql/types';
+import { ModuleIdEnum, CallIdEnum } from './common';
+import { Token } from '../../types';
+
+const chainNumberMultiplier = new BN(1000000);
+
+export async function mapToken(
+  blockId: number,
+  callId: CallIdEnum,
+  moduleId: ModuleIdEnum,
+  params: Record<string, any>,
+  extrinsic: SubstrateExtrinsic,
+): Promise<void> {
+  if (!extrinsic.success || ![ModuleIdEnum.Asset].includes(moduleId)) {
+    return;
+  }
+
+  if (callId === CallIdEnum.CreateAsset) {
+    const {
+      name,
+      ticker,
+      assetType: type,
+      fundingRound,
+      divisible,
+      disableIu,
+      identifiers,
+    } = params;
+    const ownerDid = extrinsic.events[1].event.data[0].toString();
+    await Token.create({
+      id: ticker,
+      ticker,
+      name,
+      type,
+      fundingRound,
+      isDivisible: divisible,
+      isFrozen: false,
+      isUniquenessRequired: !disableIu,
+      documents: [],
+      identifiers: JSON.stringify(identifiers),
+      ownerDid,
+      fullAgents: [ownerDid],
+      holders: [],
+      totalSupply: '0',
+    }).save();
+  }
+
+  if (callId === CallIdEnum.RenameAsset) {
+    const { ticker, name: newName } = params;
+    const token = await Token.getByTicker(ticker);
+    if (!token) return;
+    await Token.create({
+      ...token,
+      name: newName,
+    }).save();
+  }
+
+  if (callId === CallIdEnum.SetFundingRound) {
+    const { ticker, name: newFundingRound } = params;
+    const token = await Token.getByTicker(ticker);
+    if (!token) return;
+    await Token.create({
+      ...token,
+      fundingRound: newFundingRound,
+    }).save();
+  }
+
+  if (callId === CallIdEnum.AddDocuments) {
+    const { ticker, docs } = params;
+    const token = await Token.getByTicker(ticker);
+    if (!token) return;
+    const formattedDocs = docs.map((d: any) => ({
```

Comment:
Does AddDocuments actually replace existing ones?

Reply:
Yes, every time a user makes a change to documents from UI, all existing are removed first, then all new are submitted to add extrinsic. This creates an edge case when they remove all docs, or just remove one via Polkadot, in this case there's no add call coming afterwards so docs don't get updated. We need to find and store doc ID when adding, so that we can also handle remove doc call (doc IDs are the params for remove).

Reply:
Found document id in add document call events. It now also handles remove document call.